### PR TITLE
Hide other bookmarks

### PR DESCRIPTION
**Pull request by darkdh**
fix https://github.com/brave/brave-browser/issues/5158

Need https://github.com/brave/brave-browser/pull/6344 for rename completely work

This PR make "Other Bookmarks" invisible and prevent user from adding bookmarks into "Other Bookmarks". Existing bookmarks in "Other Bookmarks" will be moved under the same name folder under "Bookmarks"(Bookmarks Bar).
We also rename "Bookmarks Bar" to "Bookmarks" in `Bookmarks Manager`, `[Always] Show Bookmarks Bar` in mac menu, menu and setting page

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- [ ] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [ ] Ran `git rebase master` (if needed).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [ ] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [x] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protection-Mode
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

### Bookmarks Bar to Bookmarks

1. Check `chrome://bookmarks/` Screen Shot 2019-10-09 at 16 47 29
2. Check main menu on Mac Screen Shot 2019-10-09 at 16 49 34
3. Check menu on all platforms Screen Shot 2019-10-09 at 16 50 19
4. Check `chrome://settings/?search=bookmarks` Screen Shot 2019-10-09 at 16 51 13

### Hide Other Bookmarks with sync disabled
1. Open old Brave in fresh profile
2. Put more than one bookmarks into "Other Bookmarks" and one bookmark into "Bookmarks Bar"
3. Check these places that it contains "Other Bookmarks"
3.1 Bookmarks Bar Screen Shot 2019-10-09 at 17 14 49
3.2 Bookmarks Manager Screen Shot 2019-10-09 at 17 15 41
3.3 Main menu on Mac Screen Shot 2019-10-09 at 17 17 04
3.4 Menu on all platforms Screen Shot 2019-10-09 at 17 18 52
3.5 Bookmarks Icon  Screen Shot 2019-10-09 at 17 19 48
3.6 Edit Bookmark Dialog  Screen Shot 2019-10-09 at 17 48 38
4. Close the old Brave and launch new Brave with the same profile
5. All existing bookmarks in "Other Bookmarks" will be moved to the same name folder at the end of "Bookmarks Bar"
6. Check all places in step 3, user should have no way to see "Other Bookmarks" or add any bookmarks into it
 
### Hide Other Bookmarks with sync enabled
1. Open old Brave in two desktop devices (A, B) 
2. Put more than one bookmarks into "Other Bookmarks" and one bookmark into "Bookmarks Bar"
3. Establish sync chain between device A and B
4. Wait until both devices displaying same bookmarks
5. Close the old Brave and launch new Brave with the same profile on device A
6. All existing bookmarks in "Other Bookmarks" will be moved to the same name folder at the end of "Bookmarks Bar" on device A
7. Check all places in step 3 of previous section, user should have no way to see "Other Bookmarks" or add any bookmarks into it on device A
8. Wait for a while, device B should receive sync update and check step 6 on device B
9. Relaunch device B with new Brave and check step 7 on device B

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

File: chromium_src/components/bookmarks/browser/bookmark_model_unittest.cc [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+  // Even other node has a child, it remains invisible
+  TestNode node_a;
+  PopulateNodeFromString("A", &node_a);
+  PopulateBookmarkNode(&node_a, model_.get(), model_->other_node());
+  EXPECT_TRUE(model_->other_node()->children().front()->IsVisible());
+  EXPECT_FALSE(model_->other_node()->IsVisible());
+}
+
+// Tests that adding a URL to a folder updates the last modified time.
```
Comment (AlexeyBarabash) [inline chromium_src/components/bookmarks/browser/bookmark_model_unittest.cc]: Why the comment says about `the last modified time`?

File: chromium_src/components/bookmarks/browser/bookmark_model_unittest.cc [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+#define ParentForNewNodes DISABLED_ParentForNewNodes
+#define MostRecentlyModifiedFolders DISABLED_MostRecentlyModifiedFolders
+#include "../../../../../components/bookmarks/browser/bookmark_model_unittest.cc"
+#undef MostRecentlyModifiedFolders
+#undef ParentForNewNodes
+#undef NodeVisibility
+namespace bookmarks {
+
+TEST_F(BookmarkModelTest, BraveNodeVisibility) {
+  // Mobile node invisible by default
```
[the remark this answers is not in the document]
Comment (darkdh) [reply chromium_src/components/bookmarks/browser/bookmark_model_unittest.cc]: The test is copied from upstream because we have to disable it and modify and extend cases for other_node.

File: chromium_src/components/bookmarks/browser/bookmark_model_unittest.cc (hunk 1 above)
Comment (darkdh) [reply chromium_src/components/bookmarks/browser/bookmark_model_unittest.cc]: ditto

File: chromium_src/components/bookmarks/browser/bookmark_node.cc [hunk 3]
```diff
@@ -0,0 +1,11 @@
+/* Copyright 2019 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+// other_node should remain hidden
+#define BRAVE_IS_VISIBLE    \
+  if (type() == OTHER_NODE) \
```
Comment (bridiver) [inline chromium_src/components/bookmarks/browser/bookmark_node.cc]: can't we use `ChromeBookmarkClient::IsPermanentNodeVisible`? That's how MOBILE is displayed on android/ios and hidden on desktop

Comment (bridiver) [reply chromium_src/components/bookmarks/browser/bookmark_node.cc]: meaning subclass `ChromeBookmarkClient` if we haven't already

File: chromium_src/components/bookmarks/browser/bookmark_model.cc [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "../../../../../components/bookmarks/browser/bookmark_model.cc"
+
+namespace bookmarks {
+
+// Move bookmarks under "Other Bookmarks" permanent node to a same name folder
+// at the end of "Bookmark Bar" permanent node
+void BookmarkModel::MigrateOtherNode() {
```
Comment (bridiver) [inline chromium_src/components/bookmarks/browser/bookmark_model.cc]: are any of these calls private? Can't we just pass the BookmarkModel in to a method somewhere else?

File: patches/chrome-browser-ui-bookmarks-recently_used_folders_combo_model.cc.patch [hunk 5]
```diff
@@ -0,0 +1,12 @@
+diff --git a/chrome/browser/ui/bookmarks/recently_used_folders_combo_model.cc b/chrome/browser/ui/bookmarks/recently_used_folders_combo_model.cc
+index 369c1c2874ecb8dc5f589282f341c882f24329d5..c8e2eabd66a2d646b0a26347003e15a746dbcf5c 100644
+--- a/chrome/browser/ui/bookmarks/recently_used_folders_combo_model.cc
++++ b/chrome/browser/ui/bookmarks/recently_used_folders_combo_model.cc
+@@ -82,6 +82,7 @@ RecentlyUsedFoldersComboModel::RecentlyUsedFoldersComboModel(
+ 
+   // And put the bookmark bar and other nodes at the end of the list.
+   items_.push_back(Item(model->bookmark_bar_node(), Item::TYPE_NODE));
++  BRAVE_RECENTLY_USED_FOLDERS_COMBO_MODEL
```
Comment (bridiver) [inline patches/chrome-browser-ui-bookmarks-recently_used_folders_combo_model.cc.patch]: can we subclass this with an override in `BookmarkBubbleView` and then call
`RemoveNode(model->other_node());` in the constructor?

File: patches/chrome-browser-ui-cocoa-bookmarks-bookmark_menu_bridge.mm.patch [hunk 6]
```diff
@@ -0,0 +1,13 @@
+diff --git a/chrome/browser/ui/cocoa/bookmarks/bookmark_menu_bridge.mm b/chrome/browser/ui/cocoa/bookmarks/bookmark_menu_bridge.mm
+index eed94ff1c8ef21522eadc73b4e976f53ed94dedd..acc16976767968dfd549006748c2713b6cbde995 100644
+--- a/chrome/browser/ui/cocoa/bookmarks/bookmark_menu_bridge.mm
++++ b/chrome/browser/ui/cocoa/bookmarks/bookmark_menu_bridge.mm
+@@ -126,7 +126,7 @@ void BookmarkMenuBridge::BuildRootMenu() {
+ 
+   // If the "Other Bookmarks" folder has any content, make a submenu for it and
+   // fill it in.
+-  if (!model->other_node()->children().empty()) {
++  BRAVE_BUILD_ROOT_MENU
```
Comment (bridiver) [inline patches/chrome-browser-ui-cocoa-bookmarks-bookmark_menu_bridge.mm.patch]: why is this needed? The other node will always be empty now, right?

File: patches/chrome-browser-ui-views-bookmarks-bookmark_bar_view.cc.patch [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+   other_bookmarks_button_->SetEnabled(false);
++  BRAVE_INIT
+   AddChildView(other_bookmarks_button_);
+ 
+   profile_pref_registrar_.Init(browser_->profile()->GetPrefs());
+@@ -1938,6 +1939,7 @@ bool BookmarkBarView::UpdateOtherAndManagedButtonsVisibility() {
+   bool has_other_children = !model_->other_node()->children().empty();
+   bool update_other =
+       has_other_children != other_bookmarks_button_->GetVisible();
++  BRAVE_UPDATE_OTHER_AND_MANAGED_BUTTONS_VISIBILITY
```
Comment (bridiver) [inline patches/chrome-browser-ui-views-bookmarks-bookmark_bar_view.cc.patch]: `update_other` should already be false since other node should always be empty and GetVisible is false

Comment (bridiver) [inline patches/chrome-browser-ui-views-bookmarks-bookmark_bar_view.cc.patch]: is this needed since `has_other_children` is false so it will set the visibility of the button to false here? https://github.com/brave/brave-core/pull/3620/files#diff-b3cabb3a3f40d70b6d97898b3f71b98fR19 so this is also maybe not needed? https://github.com/brave/brave-core/pull/3620/files#diff-b3cabb3a3f40d70b6d97898b3f71b98fR9

File: chromium_src/components/bookmarks/browser/bookmark_node.cc (hunk 3 above)
Comment (bridiver) [reply chromium_src/components/bookmarks/browser/bookmark_node.cc]: wait, you already have `BraveBookmarkClient::IsPermanentNodeVisible` so why do we need this?

Comment (darkdh) [reply chromium_src/components/bookmarks/browser/bookmark_node.cc]: the patch was originally `return visible_` to get rid of `!children().empty()`.
Then I pulled upstream unit test to run and expect it to be just invisible without override `TestBookmarkClient`

File: patches/chrome-browser-ui-cocoa-bookmarks-bookmark_menu_bridge.mm.patch (hunk 6 above)
Comment (darkdh) [reply patches/chrome-browser-ui-cocoa-bookmarks-bookmark_menu_bridge.mm.patch]: not always, chromium sync uses it as a temporary place to put foster children
https://cs.chromium.org/chromium/src/components/sync_bookmarks/bookmark_change_processor.cc?l=619&rcl=74cefac725837520ab3d553eb2fa531e5ee02f63

File: chromium_src/components/bookmarks/browser/bookmark_node.cc (hunk 3 above)
Comment (darkdh) [reply chromium_src/components/bookmarks/browser/bookmark_node.cc]: The reason that other node is not always empty listed here
https://github.com/brave/brave-core/pull/3620/files#r335092116

File: patches/chrome-browser-ui-views-bookmarks-bookmark_bar_view.cc.patch (hunk 7 above)
Comment (darkdh) [reply patches/chrome-browser-ui-views-bookmarks-bookmark_bar_view.cc.patch]: `has_other_children` will not always be true https://github.com/brave/brave-core/pull/3620/files#r335092116
and 
https://github.com/brave/brave-core/pull/3620/files#diff-b3cabb3a3f40d70b6d97898b3f71b98fR9
is to prevent it from drawing separator (`BookmarkBarView::UpdateBookmarksSeparatorVisibility`)

Comment (bridiver) [reply patches/chrome-browser-ui-views-bookmarks-bookmark_bar_view.cc.patch]: no, I'm saying `has_other_children` will always be `false` so `other_bookmarks_button_->SetVisible(has_other_children);` will always be set to `false`

File: patches/chrome-browser-ui-bookmarks-recently_used_folders_combo_model.cc.patch (hunk 5 above)
Comment (darkdh) [reply patches/chrome-browser-ui-bookmarks-recently_used_folders_combo_model.cc.patch]: addressed in 456a86f2ecfebf5f3bde81fb109c5678b1647f49

File: patches/chrome-browser-ui-views-bookmarks-bookmark_bar_view.cc.patch (hunk 7 above)
Comment (darkdh) [reply patches/chrome-browser-ui-views-bookmarks-bookmark_bar_view.cc.patch]: addressed in 456a86f2ecfebf5f3bde81fb109c5678b1647f49

File: patches/chrome-browser-ui-cocoa-bookmarks-bookmark_menu_bridge.mm.patch (hunk 6 above)
Comment (darkdh) [reply patches/chrome-browser-ui-cocoa-bookmarks-bookmark_menu_bridge.mm.patch]: addressed in 3f388ed041368c0b2b4911de083aca3733d4c775

File: chromium_src/components/bookmarks/browser/bookmark_model.cc (hunk 4 above)
Comment (darkdh) [reply chromium_src/components/bookmarks/browser/bookmark_model.cc]: addressed in 122272a4e2da3a516d56fe9672d56734a610cf59

File: chromium_src/components/bookmarks/browser/bookmark_model_unittest.cc [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+  BraveTestBookmarkClient() : TestBookmarkClient() {}
+  ~BraveTestBookmarkClient() override {}
+
+  static std::unique_ptr<BookmarkModel> CreateModel() {
+  return CreateModelWithClient(base::WrapUnique(new BraveTestBookmarkClient));
+  }
+
+ private:
+  // BookmarkClient:
+  bool IsPermanentNodeVisible(const BookmarkPermanentNode* node) override {
```
Comment (bridiver) [inline chromium_src/components/bookmarks/browser/bookmark_model_unittest.cc]: It's not clear to me what we're doing here. Changing this in a test-only bookmark client doesn't seem to really validate anything useful for us in the test?

File: chromium_src/components/bookmarks/browser/bookmark_node.cc (hunk 3 above)
Comment (darkdh) [reply chromium_src/components/bookmarks/browser/bookmark_node.cc]: removed in 3f388ed

File: chromium_src/components/bookmarks/browser/bookmark_model_unittest.cc (hunk 8 above)
Comment (darkdh) [reply chromium_src/components/bookmarks/browser/bookmark_model_unittest.cc]: addressed in 0f076bb8ef52f06fa76ddaecf2b2760cf6f577ca
